onnxruntime/core/optimizer/reshape_fusion.cc

```diff
@@ -72,6 +79,136 @@ bool ReshapeFusion::Fuse_Subgraph2(Graph& graph, const NodeArg& root_input, cons
 
   return false;
 }
+
+/**
+ * Find the subgraph that matches [root] -> Shape -> Slice -> Squeeze. Check the inputs of slice 
+ * to make sure the graph produces output with exactly one element.
+ */
+bool ReshapeFusion::Match_One_Element_Output_Subgraph_2(Graph& graph, const NodeArg& root_input, const Node& cur_node,
+                                                        int index, const logging::Logger& logger) {
+  std::vector<graph_utils::EdgeEndToMatch> parent_path{
+      {0, index, "Squeeze", {1, 11}, kOnnxDomain},
+      {0, 0, "Slice", {1, 11}, kOnnxDomain},
+      {0, 0, "Shape", {1}, kOnnxDomain}};
+  std::vector<const Node::EdgeEnd*> edges;
+  if (graph_utils::FindPath(cur_node, true, parent_path, edges, logger)) {
+    const Node& slice = edges[1]->GetNode();
+    const Node& shape = edges[2]->GetNode();
+
+    const NodeArg& shape_input = *(shape.InputDefs()[0]);
+    if (shape_input.Name() != root_input.Name()) {
+      return false;
+    }
+
+    // Check if Slice op slices 1d array (result of shape) to one element.
+    std::vector<int64_t> slice_inputs;
+    if (optimizer_utils::AppendTensorFromInitializer(graph, *(slice.InputDefs()[1]), slice_inputs, true) &&
+        optimizer_utils::AppendTensorFromInitializer(graph, *(slice.InputDefs()[2]), slice_inputs, true)) {
+      const int64_t slice_start = slice_inputs[0];
+      const int64_t slice_end = slice_inputs[1];
+      if (!(slice_end >= INT_MAX && slice_start == -1) && abs(slice_end - slice_start) != 1) {
+        return false;
+      }
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/**
+ * Check if the i-th input of the current node contains exactly one number.
+ * An input has exactly one element if it's a constant value or it has an inferred shape of [1].
+ */
+bool ReshapeFusion::Is_One_Element_Input(Graph& graph, const Node& cur_node, int index) {
+  const NodeArg* cur_node_arg = cur_node.InputDefs()[index];
+  // Check if the i-th argument is constant, if it is then it's an eligible one element input.
+  if (graph_utils::NodeArgIsConstant(graph, *cur_node_arg)) {
```

Comment:
Sorry for the naive question - But if a NodeArg is a constant, does it guarantee that it has only one element ?

Reply:
Yea I think it can still be multiple elements and we still need to check the shape. Removed this condition.